ancv/web/server.py

```diff
@@ -258,3 +262,174 @@ def server_timing_header(timings: dict[str, timedelta]) -> str:
+                    status=HTTPStatus.NOT_FOUND,
+                )
+            content = await response.text()
+            try:
+                resume_data = json.loads(content)
+                return ResumeSchema(**resume_data)
+            except json.JSONDecodeError:
+                LOGGER.error("Invalid JSON format in resume data")
+                return web.Response(
+                    text="Invalid JSON format in resume data",
+                    status=HTTPStatus.BAD_REQUEST,
+                )
+
+    def render(self, resume_data: ResumeSchema) -> str | web.Response:
+        """Renders resume data into a formatted template string.
+
+        Args:
+            resume_data: The resume data dictionary to render
+
+        Returns:
+            str: The successfully rendered resume template
+            web.Response: Error response when rendering fails
+
+        Raises:
+            ResumeConfigError: When resume data doesn't match expected schema
+            ValueError: When template rendering fails
+        """
+        try:
+            template = Template.from_model_config(resume_data)
+            rendered = template.render()
+            if isinstance(rendered, str):
+                return rendered
+            return web.Response(text="Template rendering failed")
+        except ResumeConfigError as exc:
+            return web.Response(text=str(exc))
+
+    async def root(self, request: web.Request) -> web.Response:
+        """The root endpoint, returning the rendered template with periodic refresh.
+
+        Implements a caching mechanism that refreshes the resume data at configured intervals.
+        Uses monotonic time to ensure reliable cache invalidation. Falls back to cached version
+        if refresh fails.
+
+        Args:
+            request: The incoming web request containing the client session
+
+        Returns:
+            web.Response: Contains either:
+                - Fresh or cached rendered template as text
+                - Error message with SERVICE_UNAVAILABLE status when no cache exists
+
+        Note:
+            Cache refresh occurs when:
+            - No cache exists
+            - No previous fetch timestamp exists
+            - Refresh interval has elapsed since last fetch
+        """
+        log = LOGGER.bind(request=request)
+        session: ClientSession = request.app["client_session"]
+
+        current_time = time.monotonic()
+        should_refresh = (
+            not self.cache
+            or (current_time - self.last_fetch) > self.refresh_interval.total_seconds()
+        )
+
+        if should_refresh:
+            log.debug("Fetching fresh resume data.")
+            try:
+                resume_data = await self.fetch(session)
+                if isinstance(resume_data, web.Response):
+                    return resume_data
+                rendered = self.render(resume_data)
+                if isinstance(rendered, web.Response):
```

Comment:
Let's get rid of all `isinstance` checks! What this effectively does is **mix** error handling: now we have a `typing.Union` of `str | Response`, **and** we have some normal Python exceptions as well, and have to handle both.

I think a better design is for `render` to only return `str` and raise in _all_ other cases. While exceptions are really annoying to work with, it's more in line with "normal Python" and importantly it's only one error mechanism.

For that to work, you could make your own error

```python
class SomeException(Exception):
    """Raised when ..."""
```

and `raise` that manually. You could also inherit from `ValueError` or whatever if you think that's more suitable (don't think it's necessary).

And finally, you could give it a payload error (OR create multiple new exception types, each without a payload... both should work). Then, in this top-level `root` function, we just do:

```python
try:
    x = y()
except this:
   pass
except that:
   pass
```

Like you're already doing, just more of it 🙂 A final result shouldn't need `isinstance` checks.